File: Domain/Models/User/UserValidator.cs
using System.Net.Mail;
using System.Text.RegularExpressions;
using Domain.Config;

namespace Domain.Models;

public class UserValidator
{
  private readonly UserValidatorConfig _config;
  private List<ResponseError> _errors = new();

  public UserValidator()
  {
    _config = new()
    {
      DisplayNameMinChars = 1,
      DisplayNameMaxChars = 16,
      TagChars = 6,
      PasswordMinChars = 8,
      PasswordMaxChars = 32
    };
  }

  public UserValidator WithEmailAddress(string emailAddress)
  {
    if (emailAddress.Contains(' '))
    {
      _errors.Add(new(){ Field = "EmailAddress", StatusCode = 422, Message = "Email Address cannot contain spaces." });
    }
    try
    {
      MailAddress mailAddress = new(emailAddress);
    }
    catch
    {
      _errors.Add(new(){ Field = "EmailAddress", StatusCode = 422, Message = "Email Address format is invalid." });
    }
    return this;
  }
  
  public UserValidator WithDisplayName(string displayName)
  {
    if (displayName.Length < _config.DisplayNameMinChars)
    {
      _errors.Add(new ResponseError()
      {
        Field = "DisplayName",
        StatusCode = 422,
        Message = "Display name is too short."
      });
    }

    if (displayName.Length > _config.DisplayNameMaxChars)
    {
      _errors.Add(new ResponseError()
      {
        Field = "DisplayName",
        StatusCode = 422,
        Message = "Display name is too long."
      });
    }

    string pattern = "^(?!.*__)[a-zA-Z][a-zA-Z0-9_]*$";
    bool doesMatch = Regex.IsMatch(displayName, pattern);
    if (doesMatch == false)
    {
      _errors.Add(new ResponseError()
      {
        Field = "DisplayName",
        StatusCode = 422,
        Message = "Display name contains invalid characters. Must begin with a letter; and contain only letters, numbers, and underscores."
      });
    }

    return this;
  }

  public UserValidator WithTag(string tag)
  {  
    if (tag.Length != _config.TagChars)
    {
      _errors.Add(new ResponseError()
      {
        Field = "Tag",
        StatusCode = 422,
        Message = "Tag must be exactly 6 characters."
      });
    }

    string pattern = "^[a-zA-Z0-9]*$";
    bool doesMatch = Regex.IsMatch(tag, pattern);
    if (doesMatch == false)
    {
      _errors.Add(new ResponseError()
      {
        Field = "Tag",
        StatusCode = 422,
        Message = "Tag contains invalid characters. Must contain only letters and numbers."
      });
    }

    return this;
  }

  public UserValidator WithPassword(string password, string? repeatPassword = null)
  {
    if (repeatPassword is not null && password != repeatPassword) _errors.Add(new(){ Field = "Password", StatusCode = 422, Message = "Password and Repeat Password do not match."});
    if (password.Length < _config.PasswordMinChars) _errors.Add(new(){ Field = "Password", StatusCode = 422, Message = "Password is too short." });
    if (password.Length > _config.PasswordMaxChars) _errors.Add(new(){ Field = "Password", StatusCode = 422, Message = "Password is too long." });
    return this;
  }

  public Response Validate()
  {
    return new Response()
    {
      WasSuccess = _errors.Count == 0,
      Errors = _errors
    };
  }
}

File: Homepage/src/api/helpers/userValidator/userValidator.ts
import config from "../../../config";
import { IApiError } from "../../../types/api/apiError";
import { IApiResponse } from "../../../types/api/apiResponse";

import * as EmailValidator from 'email-validator';

export default class UserValidator {
  errors: IApiError[] = [];

  withEmailAddress(emailAddress: string): UserValidator {
    if (EmailValidator.validate(emailAddress) === false) {
      const error: IApiError = {
        field: "emailAddress",
        statusCode: 422,
        errorMessage: "Email Address format is invalid."
      }
      this.errors.push(error);
    }
    return this;
  }

  withPassword(password: string, repeatPassword?: string | undefined): UserValidator {
    if (repeatPassword != null && password !== repeatPassword) {
      const error: IApiError = {
        field: "repeatPassword",
        statusCode: 422,
        errorMessage: "Password is not the same in both fields."
      }
      this.errors.push(error);
    }
    if (password.length < config.registration.newUserValidation.passwordMinChars) {
      const error: IApiError = {
        field: "password",
        statusCode: 422,
        errorMessage: "Password is too short."
      }
      this.errors.push(error);
    }
    if (password.length > config.registration.newUserValidation.passwordMaxChars) {
      const error: IApiError = {
        field: "password",
        statusCode: 422,
        errorMessage: "Password is too long."
      }
      this.errors.push(error);
    }

    return this;
  }

  withDisplayName(displayName: string): UserValidator {
    if (displayName.length < config.registration.newUserValidation.displayNameMinChars) {
      const error: IApiError = {
        field: "displayName",
        statusCode: 422,
        errorMessage: "Display name is too short."
      }
      this.errors.push(error);
    }
    if (displayName.length > config.registration.newUserValidation.displayNameMaxChars) {
      const error: IApiError = {
        field: "displayName",
        statusCode: 422,
        errorMessage: "Display name is too long."
      }
      this.errors.push(error);
    }
    return this;
  }

  withTag(tag: string): UserValidator {
    if (tag.length !== config.registration.newUserValidation.tagChars) {
      const error: IApiError = {
        field: "tag",
        statusCode: 422,
        errorMessage: "Tag must be exactly 6 characters."
      }
      this.errors.push(error);
    }
    return this;
  }

  validate(): IApiResponse {
    return {
      wasSuccess: this.errors.length === 0,
      errors: this.errors
    }
  }
}